# Plot heatmaps of genome-wide stats

Plots heatmaps of the density of specified genomic elements along the chromosomes.

## Binning the elements

```sh
$ python3 cairo_genome_stats_plots/bin_genome_stats.py -h

  bin_genome_stats.py started on 2025-03-26 13:43:54.
  usage: bin_genome_stats.py [-h] -f FAI -b IN_BED [-n BASENAME] [-o OUT_DIR] 
                             [-s WIN_SIZE] [-t WIN_STEP] [-m MIN_LEN] [-p MIN_SPAN]
  
  options:
    -h, --help            show this help message and exit
    -f, --fai FAI         (str) Path to genome index in FAI format.
    -b, --in-bed IN_BED   (str) Path to input BED file describing the windows to be 
                          tallied.
    -n, --basename BASENAME
                          (str) Basename of output files [default=BED Basename].
    -o, --out-dir OUT_DIR
                          (str) Path to output directory [default=.].
    -s, --win-size WIN_SIZE
                          (int/float) Size of windows in bp [default 100,000].
    -t, --win-step WIN_STEP
                          (int/float) Step of windows in bp [default 10,000].
    -m, --min-len MIN_LEN
                          (int/float) Minimum chromosome size in bp [default 
                          1,000,000]
    -p, --min-span MIN_SPAN
                          (int/float) Minimum genomic span in bp required to keep an 
                          element from the input BED file [default=1].
```

### Inputs

Genome FASTA index, as generated by `samtools faidx <genome.fa>` 
([docs](http://www.htslib.org/doc/samtools-faidx.html)).

```sh
chr01   73382502   7           60   61
chr02   65036697   74605558    60   61
chr03   63127771   140726207   60   61
chr06   60469010   204906115   60   61
chr04   56595772   266382949   60   61
chr08   56204502   323921991   60   61
chr07   56023676   381063242   60   61
chr05   55627136   438020653   60   61
chr09   55004135   494574915   60   61
chr10   53367998   550495793   60   61
```

The span of the regions of interest in BED 
[format](https://genome.ucsc.edu/FAQ/FAQformat.html).

```
chr01   53100   53391
chr01   60254   60418
chr01   63590   63767
chr01   70591   70702
chr01   73652   73765
chr01   74530   74921
chr01   85623   87679
chr01   88104   88308
chr01   95152   95296
chr01   95709   95866
```

Note that only the first three columns are required (`chrom`, `chromStart`, and `chromEnd`), 
following the BED specification. The additional columns can contain additional information, 
but they are not used by the software. If the objective is to generate the window statistics 
for different genetic elements (e.g., genes vs repeat elements), the software must be run 
separately.

For example, a BED file of the location of annotated genes can be extracted from a GFF/GTF 
file using the following command:

```sh
$ cat genome.gff | \
    awk '$3 == "gene" {print $1, $4-1, $5}' | \
    tr ' ' '\t' > genes.bed
```

### Usage example

```sh
python3 bin_genome_stats.py \
    --fai ./genome.fa.fai \     # Path to genome FAI
    --in-bed ./genes.bed \      # Path to input BED
    --basename genome_genes \   # Basename of output files
    --out-dir ./output \        # Path to output directory
    --min-len 1000000 \         # Calulate for scaffolds larger than 1 Mbp
    --win-size 100000 \         # Average over 100 Kbp windows
    --win-step 50000 \          # Calculate windows every 50 Kbp
    --min-span 10               # Include only input elements larger than 10 bp
```

Here, we will calculate the windowed average of the density of annotated protein-coding 
genes in a target genome. We used the BED file generated above to specify the location of 
these genes using the `--in-bed` option. We specify the size and number of chromosomes by 
providing a genome index in FAI format (`--fai`). We want our output files to be saved to 
a specific directory (`--out-dir`), using a provided basename (`--basename`) to name the 
generated files.

We also provide other option describing the length and distribution of the windows. First,
we want to calculate results only for input sequences larger than 1 Mbp (`--min-len`). The 
windows will be 100 Kbp in length (`--win-size`), and we will calculate windows every 50 
Kbp (`--win-step`). Since the step of the windows is smaller than the window size, adjacent 
windows will have some amount of overlaping sites. Setting the step of windows equal to 
their size will result in non-overlapping windows. Lastly, will filter out the input BED,
skipping over any input intervals smaller than 10 bp (`--min-span`).

### Output

After running with the example above, the script will generate an output labelled 
`genome_genes.binned_genome_stats.tsv` with the following structure:

```sh
#Chrom  StartBP  EndBP    MidBP    ElementsN  ElementsAdj  PropSites   PropSitesAdj
chr01   0        450000   225000   3          0.87682672   0.27176667  2.41027714
chr01   150000   600000   375000   1          0.29227557   0.01633778  0.14489846
chr01   300000   750000   525000   6          1.75365344   0.19475778  1.72729137
chr01   450000   900000   675000   5          1.46137787   0.07738667  0.68633624
chr01   600000   1050000  825000   3          0.87682672   0.06501556  0.57661784
chr01   750000   1200000  975000   1          0.29227557   0.04927333  0.43700131
chr01   900000   1350000  1125000  0          0.00000000   0.00000000  0.00000000
chr01   1050000  1500000  1275000  5          1.46137787   0.35131778  3.11580967
chr01   1200000  1650000  1425000  4          1.16910230   0.24542667  2.17666975
```

Info on columns:

| Column Header   | Explanation |
| --------------- | ----------- |
| `Chrom`         | Name of the chromosome/scaffold sequence. |
| `StartBP`       | Start coordinate of the window in basepairs. |
| `EndBP`         | End coordinate of the window in basepairs. |
| `MidBP`         | Midpoint coordinate of the window in basepairs. |
| `ElementsN`     | Number of elements seen in the window. |
| `ElementsAdj`<sup>1</sup> | Adjusted number of elements in the window, according to the genome-wide average. |
| `PropSites`<sup>2</sup> | Proportion of sites in the window corresponding to the target elements. Sites belonging to overlapping elements are counted once. |
| `PropSitesAdj`<sup>1</sup> | Adjusted Proportion of sites in the window corresponding to the target elements, according to the genome-wide average. |

<sup>1</sup>For standardization purposes, the adjusted values 
correspond to the value of the window divided by the genome-wide mean, i.e., the 
`ElementsAdj` describes the number of elements of interest observed in a window,
scaled by the genome-wide average. Values greater than 1 describe sites with 
higher than average values, while values smaller than 1 describe window with smaller 
than average values. For example, a `ElementsAdj` of 2 indicates that the given 
window contains twice as elements than the genome-wide average.

<sup>2</sup>The `PropSites` column describes the proportion of sites in a window 
corresponding to the elements of interest. In other words, if in the span of a 
100 Kbp window, 30 Kbp of those sites are in the span of an element of interest 
(e.g., genes), then the window will have a `PropSites` value of 0.3. Similar to 
`ElementsAdj`, this value is then scaled by the genome-wide average in `PropSitesAdj`.

## Making the plots

Requires [PyCairo](https://pycairo.readthedocs.io/en/latest/).

```sh
$ ./cairo_plot_genome_stats.py -h
usage: cairo_plot_genome_stats.py [-h] --chroms CHROMS
                                  --fai FAI --reps-tsv
                                  REPS_TSV --prot-tsv
                                  PROT_TSV
                                  [--out-dir OUT_DIR]
                                  --name NAME
                                  [--img-height IMG_HEIGHT]
                                  [--img-width IMG_WIDTH]
                                  [--img-format IMG_FORMAT]
                                  [--scale SCALE]
                                  [--step STEP]
                                  [--min-len MIN_LEN]

options:
  -h, --help            show this help message and exit
  --chroms CHROMS       Chromosome order list
  --fai FAI             Genome FASTA index
  --reps-tsv REPS_TSV   Repeat window tally table
  --prot-tsv PROT_TSV   Protein window tally table
  --out-dir OUT_DIR     Output directory
  --name NAME           Name of the dataset
  --img-height IMG_HEIGHT
                        Image height in pixels
  --img-width IMG_WIDTH
                        Image width in pixels
  --img-format IMG_FORMAT
                        Image output format
  --scale SCALE         Scale chromosome lengths to this
                        value
  --step STEP           Steps for size tick marks
  --min-len MIN_LEN     Minimum length of chromosomes
```

### Inputs

Chromosome order list: Text file containing the chromosome IDs (one per line) in the order they will be plotted.

```sh
chr01
chr02
chr03
chr06
chr04
chr08
chr07
chr05
chr09
chr10
```

## Some examples

Figure 2B and 2C ([link](https://academic.oup.com/view-large/figure/397327756/msad029f3.tif)) from [Rivera-Colon et al. 2023](https://doi.org/10.1093/molbev/msad029).

Figure 3 ([link](https://academic.oup.com/view-large/figure/499727840/jkae267f3.jpg)) from [Rayamajhi et al. 2024](https://doi.org/10.1093/g3journal/jkae267).

TODO: Add some images.

## Author

Angel G. Rivera-Colon  
Institute of Ecology and Evolution  
University of Oregon
